### PR TITLE
bug(cli):fix kubectl rollout status not recoginze resource namespace

### LINKED ISSUE #42364: "kubectl rollout status -f resource.yml" doesn't use namespace from resource.yml
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): rollout status namespace

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug Report

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:40:50Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Macbook Pro 2015
- **OS** (e.g. from /etc/os-release): Mac OSX Sierra
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:

1. Using `kubectl apply -f resource.yml`, where resource.yml does the following:
  - Describes a Deployment object
  - Sets a non-default namespace in the metadata
2. Run `kubectl rollout status -f resource.yml`. An error message says the resource was not found
3. Run `kubectl rollout status -f resource.yml -f <customNamespace>`. The response shows the status of the rollout.

**What you expected to happen**:
I expected `kubectl rollout status -f resource.yml` to work as well by using the namespace that is set inside resource.yml for the rollout status request.

**How to reproduce it** (as minimally and precisely as possible):
1. Run `kubectl create namespace reprns`
2. Create the following file as nginx.yml (taken from the [nginx tutorial docs](https://raw.githubusercontent.com/kubernetes/kubernetes.github.io/master/docs/tutorials/stateless-application/deployment.yaml) but updated to add the namespace):
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: reprns
spec:
  replicas: 2 # tells deployment to run 2 pods matching the template
  template: # create pods using pod definition in this template
    metadata:
      # unlike pod-nginx.yaml, the name is not included in the meta data as a unique name is
      # generated from the deployment name
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80
```
3. Run `kubectrl apply -f nginx.yml`
4. Run `kubectl rollout status -f nginx.yml`
  - Returns `Error from server (NotFound): deployments.extensions "nginx-deployment" not found`
5. Run `kubectl rollout status -f nginx.yaml -n reprns`
  - Returns `deployment "nginx-deployment" successfully rolled out`

**Anything else we need to know**:
We've seen this issue on 1.5.1 as well, and also on Ubuntu. The recreate instructions above were performed on a fresh minikube setup.

### PR DESCRIPTION
**Pull request by zjj2wry**
**What this PR does / why we need it**:
Fixes #42364

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
cc @kubernetes/sig-cli-pr-reviews

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/make-rules/test-cmd-util.sh [hunk 1]
```diff
@@ -2824,6 +2824,18 @@ run_deployment_tests() {
...
+  kube::test::get_object_assert deployment "{{range.items}}{{$id_field}}:{{end}}" ''
+  # create namespace 'my-namespace'
+  kubectl create namespace my-namespace "${kube_flags[@]}"
+  kubectl config set-context "${CONTEXT}" --namespace="my-namespace"
+  kube::test::get_object_assert 'namespace my-namespace' "{{$id_field}}" 'my-namespace'
+  kubectl create -f hack/testdata/deployment-rollout-status.yaml "${kube_flags[@]}"
+  kubectl rollout status -f hack/testdata/deployment-rollout-status.yaml "${kube_flags[@]}"
+  # Clean up
+  kubectl delete deployment nginx "${kube_flags[@]}"
+
```
Comment (zjj2wry) [inline hack/make-rules/test-cmd-util.sh]: @mengqiy i am no idea about this test, can help me fix the test fail? appreciate

**Comment by mengqiy** [comment]
This supersedes #42540.

File: hack/make-rules/test-cmd-util.sh (hunk 1 above)
Comment (mengqiy) [reply hack/make-rules/test-cmd-util.sh]: I think test failure is caused by `run_deployment_tests` has called `create_and_use_new_namespace` at the beginning of the test. 
So I believe there is a `--namespace` flag in `"${kube_flags[@]}"` that may override your flag specified in the file.

You may want to move your new test to a new test case name that doesn't call `create_and_use_new_namespace`.

**Comment by zjj2wry** [comment]
seems this image `gcr.io/google-containers/nginx:test-cmd` can pulled? this rollout will always pending.
```
I0928 04:22:16.284] +++ Running case: test-cmd.run_rollout_tests 
I0928 04:22:16.286] +++ working dir: /go/src/k8s.io/kubernetes
I0928 04:22:16.288] +++ command: run_rollout_tests
I0928 04:22:16.408] namespace "my-namespace" created
I0928 04:22:16.517] Context "test" modified.
I0928 04:22:16.730] deployment "nginx" created
W0928 04:22:16.831] I0928 04:22:16.734700   15456 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"my-namespace", Name:"nginx", UID:"9c4dd09a-a404-11e7-a0c2-0242ac110002", APIVersion:"extensions", ResourceVersion:"2812", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set nginx-87b94df67 to 3
W0928 04:22:16.831] I0928 04:22:16.749156   15456 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-87b94df67", UID:"9c4e84ae-a404-11e7-a0c2-0242ac110002", APIVersion:"extensions", ResourceVersion:"2813", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-87b94df67-gdq46
W0928 04:22:16.832] I0928 04:22:16.754908   15456 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-87b94df67", UID:"9c4e84ae-a404-11e7-a0c2-0242ac110002", APIVersion:"extensions", ResourceVersion:"2813", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-87b94df67-fmw42
W0928 04:22:16.832] I0928 04:22:16.754959   15456 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-87b94df67", UID:"9c4e84ae-a404-11e7-a0c2-0242ac110002", APIVersion:"extensions", ResourceVersion:"2813", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-87b94df67-56pgh
I0928 04:22:16.932] Waiting for rollout to finish: 0 of 3 updated replicas are available...

```

**Comment by zjj2wry** [comment]
```
I1010 16:22:05.376] Waiting for rollout to finish: 0 of 3 updated replicas are available...
I1010 16:52:16.866] Terminate 30914 on timeout
E1010 16:52:16.967] Build timed out
E1010 16:52:16.968] Command failed
I1010 16:52:16.968] process 30914 exited with code -15 after 59.3m
```
this has trigger `rollout status`, but rollout always pendding, so cause this ci fail.

**Comment by zjj2wry** [comment]
```
W1020 04:50:07.759] I1020 04:50:07.660639   16894 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"my-namespace", Name:"nginx", UID:"2557c0fc-b552-11e7-8366-0242ac110002", APIVersion:"extensions", ResourceVersion:"3001", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set nginx-5686cbf55b to 3
W1020 04:50:07.760] I1020 04:50:07.667761   16894 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-5686cbf55b", UID:"255860d8-b552-11e7-8366-0242ac110002", APIVersion:"extensions", ResourceVersion:"3002", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-5686cbf55b-szftc
W1020 04:50:07.760] I1020 04:50:07.670097   16894 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-5686cbf55b", UID:"255860d8-b552-11e7-8366-0242ac110002", APIVersion:"extensions", ResourceVersion:"3002", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-5686cbf55b-jjgxj
W1020 04:50:07.760] I1020 04:50:07.670900   16894 event.go:218] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"my-namespace", Name:"nginx-5686cbf55b", UID:"255860d8-b552-11e7-8366-0242ac110002", APIVersion:"extensions", ResourceVersion:"3002", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: nginx-5686cbf55b-fpfkg
I1020 04:50:07.860] Waiting for rollout to finish: 0 of 3 updated replicas are available...
```

**Comment by zjj2wry** [comment]
@liggitt Do you know this why alway pendding?

File: hack/testdata/deployment-rollout-status.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+apiVersion: extensions/v1beta1
```
Comment (dixudx) [inline hack/testdata/deployment-rollout-status.yaml]: use `hack/testdata/deployment-revision2.yaml` instead?

File: hack/make-rules/test-cmd-util.sh [hunk 3]
```diff
@@ -3107,6 +3107,23 @@ run_daemonset_tests() {
   set +o errexit
 }
 
+run_rollout_tests() {
+  set -o nounset
+  set -o errexit
+
+  ### Show the status of the rollout
+  # create namespace 'my-namespace'
```
Comment (dixudx) [inline hack/make-rules/test-cmd-util.sh]: Here we usually use `create_and_use_new_namespace` instead.

File: hack/make-rules/test-cmd-util.sh [hunk 4]
```diff
@@ -3107,6 +3107,23 @@ run_daemonset_tests() {
...
+  kubectl create namespace "my-namespace"
+  kubectl config set-context "${CONTEXT}" --namespace="my-namespace"
+  kubectl create -f hack/testdata/deployment-rollout-status.yaml
+  kubectl rollout status -f hack/testdata/deployment-rollout-status.yaml
+  # Clean up
+  kubectl delete ns my-namespace
+
+  set -o nounset
```
Comment (dixudx) [inline hack/make-rules/test-cmd-util.sh]: Wrong set here. Please change to 

```
    set +o nounset
    set +o errexit
```

File: hack/make-rules/test-cmd-util.sh (hunk 3 above)
Comment (zjj2wry) [reply hack/make-rules/test-cmd-util.sh]: i want test rollout status will recoginze namespace in my local yaml file. and my yaml have namespace `NAME`

**Comment by zjj2wry** [comment]
I1027 07:50:05.330] Waiting for rollout to finish: 0 of 1 updated replicas are available...
I1027 08:20:38.570] Terminate 19443 on timeout
E1027 08:20:38.672] Build timed out

**Comment by zjj2wry** [comment]
```
'Warning' reason: 'FailedPlacement' failed to place pod on "127.0.0.1": Node didn't have enough resource: pods, requested: 1, used: 0, capacity: 0
```
i understand pod will not running actual. remove the test and fix the bug directly.

**Comment by zjj2wry** [comment]
@liggitt this is a bug, can we let it fixes in 1.9? thanks

**Pull request merged**